MNT move an orphaned vendored file around

DimitriPapadopoulos opened the pull request:
#### Reference Issues/PRs

See also #21161 and #21165.

#### What does this implement/fix? Explain your changes.

File `scipy/misc/pilutil.py` has been removed from Scipy 1.3.0, see scipy/scipy#9325.

Move `_pilutil.py` from `sklearn/externals`, which is the location for vendored files, to `sklearn/utils`, which is a location for
locally maintained files, including orphan vendored files like this one.

#### Any other comments?

This is a copy of #21165. Something went wrong while rebasing.

DimitriPapadopoulos committed:
MNT move an orphaned vendored file around

File scipy/misc/pilutil.py has been removed from Scipy 1.3.0,
see scipy/scipy#9325.

Move _pilutil.py from sklearn/externals, which is the location
for vendored files, to sklearn/utils, which is a location for
locally maintained files, including orphan vendored files like
this one.

DimitriPapadopoulos commented:
Since this is a file for internal use, does this need to be in the changelog?

DimitriPapadopoulos committed:
MNT pass flake8 CI jobs

DimitriPapadopoulos commented:
Not sure about the Azure CI `black` job failure. Is `sklearn/externals` exempted from black, but not `sklearn/utils`?

DimitriPapadopoulos committed:
MNT pass black CI job

thomasjpfan commented:
> Not sure about the Azure CI black job failure. Is sklearn/externals exempted from black, but not sklearn/utils?

Yea. `sklearn/externals` is for vendored projects/files so we try not to change the source.  `sklearn/utils` is maintained by us, which follows `black`.

DimitriPapadopoulos commented:
@thomasjpfan Makes sense. I ran flake8 and black on the file, but some Azure CI jobs still fail.

DimitriPapadopoulos commented:
I had seen [that page](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=33042&view=logs&j=78a0bf4f-79e5-5387-94ec-13e67d216d6e&t=f1857171-4a53-55c7-3ab5-90acfe091baa), but it doesn't make sense:
* It's too verbose.
* It looks like the problem is with the `XFAIL` messages.

However, now that you've pointed me to these doctest messages, I'll try fixing them:
```
__________________ [doctest] sklearn.utils._pilutil.bytescale __________________
[gw0] linux -- Python 3.9.7 /usr/share/miniconda/envs/testvenv/bin/python
108 
109     Returns
110     -------
111     img_array : uint8 ndarray
112         The byte-scaled array.
113 
114     Examples
115     --------
116     >>> import numpy as np
117     >>> from scipy.misc import bytescale
UNEXPECTED EXCEPTION: ImportError("cannot import name 'bytescale' from 'scipy.misc' (/usr/share/miniconda/envs/testvenv/lib/python3.9/site-packages/scipy/misc/__init__.py)")
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/testvenv/lib/python3.9/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest sklearn.utils._pilutil.bytescale[1]>", line 1, in <module>
ImportError: cannot import name 'bytescale' from 'scipy.misc' (/usr/share/miniconda/envs/testvenv/lib/python3.9/site-packages/scipy/misc/__init__.py)
/home/vsts/work/1/s/sklearn/utils/_pilutil.py:117: UnexpectedException
___________________ [doctest] sklearn.utils._pilutil.imsave ____________________
[gw0] linux -- Python 3.9.7 /usr/share/miniconda/envs/testvenv/bin/python
243         Image format. If omitted, the format to use is determined from the
244         file name extension. If a file object was used instead of a file name,
245         this parameter should always be used.
246 
247     Examples
248     --------
249     Construct an array of gradient intensity values and save to file:
250 
251     >>> import numpy as np
252     >>> from scipy.misc import imsave
UNEXPECTED EXCEPTION: ImportError("cannot import name 'imsave' from 'scipy.misc' (/usr/share/miniconda/envs/testvenv/lib/python3.9/site-packages/scipy/misc/__init__.py)")
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/testvenv/lib/python3.9/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest sklearn.utils._pilutil.imsave[1]>", line 1, in <module>
ImportError: cannot import name 'imsave' from 'scipy.misc' (/usr/share/miniconda/envs/testvenv/lib/python3.9/site-packages/scipy/misc/__init__.py)
/home/vsts/work/1/s/sklearn/utils/_pilutil.py:252: UnexpectedException
```

DimitriPapadopoulos committed:
MNT pass doctest CI job

Either import the private module documented in these examples, or we
remove the examples altogether. I chose to keep the examples.

DimitriPapadopoulos commented:
It looks like doctests are OK after my last commit. Please let me know if anything else is needed.

thomasjpfan reviewed (dismissed):
Looking at this again, I'm okay with the moving the vendored file into `utils` for now.

LGTM

jjerphan reviewed (approved):
Thank you, @DimitriPapadopoulos.

I am mostly OK to move this file under `utils` for maintenance semantic reasons since it has been deprecated in Scipy. (I think having as little vendored code as possible is good for maintenance.)

IMO, it looks like the implementations can be simplified, but that's out of the scope of this PR (and it is also probably low priority as a change).

---

Edit: I would wait for a third reviewer before merging this one.

thomasjpfan commented:
After looking at this again, I prefer removing the file all together. I opened https://github.com/scikit-learn/scikit-learn/pull/22743 as an alternative, which removes `_pilutil`.

ogrisel commented:
I would also lean towards the alternative implemented in #22743.

lesteve commented:
Closing since #22743 was merged.
